2020/7/luggage/src/main.rs

```diff
@@ -0,0 +1,202 @@
+use std::hash::{Hash, Hasher};
+
+fn main() -> Result<(), std::io::Error> {
+    let args : Vec<_> = env::args().collect();
+    let input = fs::read_to_string(&args[1])?;
+    let rules = Rules::parse(&input);
+
+    println!("shiny gold bags contain: {}", rules.contains_recursive("shiny gold"));
+
+    Ok(())
+}
+
+pub struct Rules<'a> {
+    rules: HashSet<Rule<'a>>,
+}
+
+impl<'a> Rules<'a> {
+    pub fn parse(input: &'a str) -> Rules {
+        Rules {
+            rules: input
+                .split('\n')
+                .map(|line| Rule::parse(&line))
+                .collect(),
+        }
+    }
+
+    pub fn can_eventually_contain(&self, bag_name: &str) -> Option<HashSet<&'a Rule>> {
+        if let Some(r) = self.rules.iter().filter(|x| x.name == bag_name).nth(0) {
+            let mut v = HashSet::new();
+            v.insert(r);
+            v = self.can_contain_recursive(v, 0);
+            v.remove(r);
+            Some(v)
+        } else {
+            None
+        }
+    }
+
+    fn can_contain_recursive(&self, rules: HashSet<&'a Rule<'a>>, depth: usize) -> HashSet<&'a Rule> {
+        let names : Vec<&'a str> = rules.iter().map(|r| r.name).collect();
+        println!("Looking for rules that can contain {:?}", names);
+        let mut new = HashSet::new();
+        for r1 in &self.rules {
+            for r2 in &rules {
+                if r1.can_contain_directly(r2.name) {
+                    new.insert(r1);
+                }
+            }
+        }
+
+        let len = rules.len();
+        for r in rules {
+            new.insert(r);
+        }
+
+        if new.len() == len {
+            return new
+        }
+
+        self.can_contain_recursive(new, depth + 1)
+    }
+
+    pub fn contains_recursive(&self, bag_name: &str) -> usize {
+        self.contains_recursive_helper(bag_name) - 1
+    }
+
+    fn contains_recursive_helper(&self, bag_name: &str) -> usize {
+        if let Some(r) = self.rules.iter().filter(|x| x.name == bag_name).nth(0) {
+            1 + r.contains_recursive(self)
+        } else {
+            0
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Rule<'a> {
+    name: &'a str,
+    contains: HashSet<(&'a str, usize)>,
+}
+
+impl<'a> Rule<'a> {
+    pub fn parse(line: &str) -> Rule {
+        let mut spaces = Vec::new();
+        for (pos, c) in line.chars().enumerate() {
+            if c == ' ' {
+                spaces.push(pos);
+            }
+        }
+        spaces.push(line.len());
+
+        let mut contains = Vec::new();
+        if &line[spaces[3]..] == " no other bags." {
+        } else {
+            let mut space = 3;
+            while space < spaces.len() - 1 {
+                let bn = &line[spaces[space + 1] + 1..spaces[space + 3]];
+                let c = line[spaces[space] + 1..spaces[space+1]].parse::<usize>().unwrap();
+                contains.push((bn, c));
+                space += 4;
+            }
+        }
+
+        // Couldn't get collect to work below :(
```

Comment:
Strange. I would've expected this to work, at least:

```rust
let hs: HashSet<_> = contains.iter().collect();
```

Reply:
```
mismatched types

expected tuple, found reference

note: expected type `std::collections::HashSet<(&str, usize)>`
         found type `std::collections::HashSet<&(&str, usize)>`rustc(E0308)
main.rs(115, 23): expected tuple, found reference
```